packages/clerk-js/src/ui/router/PathRouter.tsx

```diff
@@ -36,8 +36,12 @@ export const PathRouter = ({ basePath, preservedParams, children }: PathRouterPr
   React.useEffect(() => {
     const effect = async () => {
       if (window.location.hash.startsWith('#/')) {
+        const url = new URL(window.location.href);
```

Comment:
@desiprisg Can we please rewrite the hash-to-path conversion logic to make it more intuitive and readable?

Reply:
@desiprisg 
Shall we try this approach? 
```ts
/**
 * Creates a new URL by merging a fragment-based URL, if found.
 * The result URL has the original and the fragment pathnames appended
 * and also includes all search params from both places.
 *
 * @example
 * Input
 * https://accounts.clerk.com/sign-in?user_param=hello#/verify/factor-one?redirect_url=/protected
 * Return value:
 * https://accounts.clerk.com/sign-in/verify/factor-one?user_param=hello&redirect_url=/protected
 */
const mergeFragmentUrl = (_url: string | URL) => {
  const url = new URL(_url);
  const hasUrlInFragment = url.hash.startsWith('#/');

  if (!hasUrlInFragment) {
    return url;
  }

  const fragmentUrl = new URL(url.hash.replace('#/', '/'), url.href);
  const mergedPathname = [url.pathname, fragmentUrl.pathname]
    .map(s => s.split('/'))
    .flat()
    .filter(Boolean)
    .join('/');

  const mergedUrl = new URL(mergedPathname, url);

  url.searchParams.forEach((val, key) => {
    mergedUrl.searchParams.set(key, val);
  });

  fragmentUrl.searchParams.forEach((val, key) => {
    mergedUrl.searchParams.set(key, val);
  });

  return mergedUrl;
};
```

We should also add the necessary unit tests.